rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+        component.
+    *   Potentially, `BaseExecutor` can extend this model to support a class
+        method `validate_component_spec()` to support user executor static
+        validation of any logic.
+*   **Cons**
+    *   `dynamic_spec_class()` style may be cumbersome.
+    *   The dynamic class cannot be shared like static class.
+
+## Appendix
+
+### Ppeline Compilation and Release
+
+The proposed `ExecutorContainerSpec` and related modification will reside in the
+TFX repository.
+
+If run with `KubeflowDagRunner`, it will be executed by `run()` method to
+compile into Argo pipeline spec. As a result, there is no need to have any
+additional code to be included inside the user’s container image. Other
+orchestrators, such as `AirflowDAGRunner`, may have to have a newer version of
+TFX SDK with the new `ExecutorContainerSpec` and implementations of
+corresponding `ComponentSpec` subclasses installed in the environment in which
+the component is executed. Nevertheless, it is no different than any other TFX
+component’s execution in which it needs to have the TFX SDKs for components
+installed on the Airflow execution environment.
+
+### Componentize a Python function, as opposed to a container image
+
+Kubeflow Pipelines SDK helps users to define a Python function and convert them
+to a container-based application as a part of the pipeline (by the
+`kfp.compiler.build_python_component()` API). In order for this to become fully
+metadata-aware component as proposed in this document, a gap still remains that
+it doesn’t help defining input- and output- of the Python function in terms of
+the typed Artifacts to be tracked in ML Metadata.
+
+This proposed container-based component could further help filling the gap to
+help declaratively configure `INPUTS`, `OUTPUTS` and `PARAMETERS` for the given
+naked Python function and componentize it. Furthermore, there is an opportunity
+to create a helper to build an image, configure a `python` command entrypoint
+from a naked Python function, and construct command line arguments under the
+hood, as a specialized subclass of it. Such helper shall eventually converge to
+the other way of implementing a
+[custom component](https://github.com/tensorflow/tfx/tree/80a80ffb0f833cf02344ac2898017cf1b73e166d/tfx/examples/custom_components)
+for TFX via a custom `Executor` class written in Python, and package it in a
+container image to release.
+
+Detailed RFC particularly on this point will follow.
+
+### Component Arhcetypes
+
+As of `tfx` 0.14, there are
+[10 known artifact types](https://github.com/tensorflow/tfx/blob/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/types/standard_artifacts.py)
+defined and used.
+
+*   ExternalArtifact
+*   Examples
+*   Schema
+*   ExampleStatistics
+*   ExampleAnomalies
+*   TransformGraph
+*   Model
+*   ModelEvalulation
+*   ModelBlessing
+*   PushedModel
+
+Based on the above known artifact types, TFX defines the following
```

Comment:
artifact type used in the following table needs some update, see inline comments

Reply:
Done